File: R/Rglpk_solve.R
## the R-ported GNU Linear Programming kit
## solve function --- C Interface

Rglpk_solve_LP <- function(obj, mat, dir, rhs, bounds = NULL, types = NULL, max = FALSE,
                           control = list(), ...)
{
    ## validate control list
    dots <- list(...)
    control[names(dots)] <- dots
    control <- .check_control_parameters( control )
    canonicalize_status <- control$canonicalize_status
    presolve <- control$presolve
    time_limit <- control$tm_limit
    verb <- control$verbose
    sensitivity_report <- isTRUE(control$sensitivity_report)

    Rglpk_call( obj = obj, mat = mat, dir = dir, rhs = rhs, bounds = bounds,
      types = types, max = max, canonicalize_status = canonicalize_status,
      presolve = presolve, time_limit = time_limit, verb = verb,
      sensitivity_report = sensitivity_report)
}

Rglpk_call <- function(obj, mat, dir, rhs, bounds, types, max, canonicalize_status,
  presolve, time_limit, verb, file = "", file_type = 0L, sensitivity_report = FALSE) {
  ## validate direction of optimization
  if(!identical( max, TRUE ) && !identical( max, FALSE ))
      stop("'Argument 'max' must be either TRUE or FALSE.")
  direction_of_optimization <- as.integer(max)

  ## match direction of constraints
  n_of_constraints <- length(dir)
  ## match relational operators to requested input
  direction_of_constraints <- match( dir, c("<", "<=", ">", ">=", "==") )

  if( any(is.na(direction_of_constraints)) )
    stop("Argument 'dir' must be either '<', '<=', '>', '>=' or '=='.")

  ## we need to verify that obj is a numeric vector
  ## FIXME: always use STMs?
  if(slam::is.simple_triplet_matrix(obj))
      obj <- as.matrix(obj)
  obj <- as.numeric(obj)
  n_of_objective_vars <- length( obj )

  constraint_matrix <- as.simple_triplet_matrix(mat)

  ## types of objective coefficients
  ## Default: "C"
  if(is.null(types))
    types <- "C"
  ## check if valid types
  if(any(is.na(match(types, c("I", "B", "C"), nomatch = NA))))
    stop("'types' must be either 'B', 'C' or 'I'.")
  ## replicate types to fit number of columns
  types <- rep(types, length.out = n_of_objective_vars)
  ## need a TRUE/FALSE integer/binary representation
  integers <- types == "I"
  binaries <- types == "B"

  ## do we have a mixed integer linear program?
  is_integer <- any( binaries | integers )

  if ( sensitivity_report & is_integer ) {
    stop("GLPK does not support sensitivity analysis report for mixed integer problems")
  }

  ## bounds of objective coefficients
  bounds <- as.glp_bounds( as.list( bounds ), n_of_objective_vars )

  ## GLPK doesn't fix the variables if the type is binary,
  ## so we just change the type of fixed binary variables integer.
  ## reported by Dirk Schumacher, patch submitted by Florian Schwendinger
  change_to_integer <- which(binaries & bounds[, 1] == 5L)
  if ( length(change_to_integer) ) {
      if ( !all(unlist(bounds[change_to_integer, 2:3]) %in% c(0L, 1L)) ) {
          stop("binary variables can only be fixed to 0L or 1L")
      }
      integers[change_to_integer] <- TRUE
      binaries[change_to_integer] <- FALSE
  }

  ## Sanity check: mat/dir/rhs
  if( !all(c(dim(mat)[ 1 ], length(rhs)) == n_of_constraints) )
      stop( "Arguments 'mat', 'dir', and/or 'rhs' not conformable." )
  ## Sanity check: mat, obj
  if( dim(mat)[ 2 ] != n_of_objective_vars )
      stop( "Arguments 'mat' and 'obj' not conformable." )

  ## file writer functionality
  if(file_type %in% 1:2){
      if( direction_of_optimization )
          obj <- -obj
      direction_of_optimization <- 0L
  }

  if (sensitivity_report) {
    write_sensitivity_report <- 1L
    fname_sensitivity_report <- tempfile()
  } else {
    write_sensitivity_report <- 0L
    fname_sensitivity_report <- ""
  }

  ## call the C interface - this actually runs the solver
  x <- glp_call_interface(obj, n_of_objective_vars, constraint_matrix$i,
                          constraint_matrix$j, constraint_matrix$v,
                          length(constraint_matrix$v),
                          rhs, direction_of_constraints, n_of_constraints,
                          is_integer,
                          integers, binaries,
                          direction_of_optimization, bounds[, 1L],
                          bounds[, 2L], bounds[, 3L], verb, presolve, time_limit,
                          file_type, file, write_sensitivity_report, fname_sensitivity_report)

  solution <- x$lp_objective_vars_values
  ## are integer variables really integers? better round values
  solution[integers | binaries] <- round( solution[integers | binaries])
  ## match status of solution
  status <- as.integer(x$lp_status)
  if(canonicalize_status) {
    ## 0 -> optimal solution (5 in GLPK) else 1
    status <- as.integer(status != 5L)
  }

  if (sensitivity_report) {
      sensitivity_report <- readLines(fname_sensitivity_report)
      file.remove(fname_sensitivity_report)
  } else {
    sensitivity_report <- NA_character_
  }

  list(optimum = sum(solution * obj), solution = solution, status = status,
       solution_dual = if( is_integer ) NA else x$lp_objective_dual_values,
       auxiliary = list(primal = x$lp_row_prim_aux,
                        dual   = if( is_integer) NA else x$lp_row_dual_aux),
       sensitivity_report = sensitivity_report)
}

## this function calls the C interface
glp_call_interface <-
function(lp_objective_coefficients, lp_n_of_objective_vars,
         lp_constraint_matrix_i, lp_constraint_matrix_j, lp_constraint_matrix_v,
         lp_n_of_values_in_constraint_matrix, lp_right_hand_side,
         lp_direction_of_constraints, lp_n_of_constraints, lp_is_integer,
         lp_objective_var_is_integer, lp_objective_var_is_binary,
         lp_direction_of_optimization,
         lp_bounds_type, lp_bounds_lower, lp_bounds_upper,
         verbose, presolve, time_limit, write_fmt, fname, write_sensitivity_report,
         fname_sensitivity_report)
{
  out <- .C(R_glp_solve,
            lp_direction_of_optimization= as.integer(lp_direction_of_optimization),
            lp_n_of_constraints         = as.integer(lp_n_of_constraints),
            lp_direction_of_constraints = as.integer(lp_direction_of_constraints),
            lp_right_hand_side          = as.double(lp_right_hand_side),
            lp_n_of_objective_vars      = as.integer(lp_n_of_objective_vars),
            lp_objective_coefficients   = as.double(lp_objective_coefficients),
            lp_objective_var_is_integer = as.integer(lp_objective_var_is_integer),
            lp_objective_var_is_binary  = as.integer(lp_objective_var_is_binary),
            lp_is_integer               = as.integer(lp_is_integer),
            lp_n_of_values_in_constraint_matrix = as.integer(lp_n_of_values_in_constraint_matrix),
            lp_constraint_matrix_i      = as.integer(lp_constraint_matrix_i),
            lp_constraint_matrix_j      = as.integer(lp_constraint_matrix_j),
            lp_constraint_matrix_values = as.double(lp_constraint_matrix_v),
            lp_bounds_type              = as.integer(lp_bounds_type),
            lp_bounds_lower             = as.double(lp_bounds_lower),
            ## lp_n_of_bounds_l            = as.integer(length(lp_lower_bounds_i)),
            lp_bounds_upper             = as.double(lp_bounds_upper),
            ## lp_n_of_bounds_u            = as.integer(length(lp_upper_bounds_i)),
            lp_optimum                  = double(1),
            lp_objective_stat           = integer(lp_n_of_objective_vars),
            lp_objective_vars_values    = double(lp_n_of_objective_vars),
            lp_objective_dual_values    = double(lp_n_of_objective_vars),
            lp_row_stat                 = integer(lp_n_of_constraints),
            lp_row_prim_aux             = double(lp_n_of_constraints),
            lp_row_dual_aux             = double(lp_n_of_constraints),
            lp_verbosity                = as.integer(verbose),
            lp_presolve                 = as.integer(presolve),
            lp_time_limit               = as.integer(time_limit),
            lp_status                   = integer(1),
            write_fmt                   = as.integer(write_fmt),
            fname                       = as.character(fname),
            write_sensitivity_report    = write_sensitivity_report,
            fname_sensitivity_report    = fname_sensitivity_report,
            NAOK = TRUE, PACKAGE = "Rglpk")
  out
}

## Convenience function for solving MILP objects
## upcoming ROI package (or for solving problems read with filereader)
.ROI_glpk_solve <- function(x, control = list()){
  if(!inherits(x, "MILP"))
    stop("'x' must be of class 'MILP'")
  if(is.null(control$verbose))
    control$verbose <- FALSE
  Rglpk_solve_LP(x$objective, x$constraints[[1]],
                 x$constraints[[2]], x$constraints[[3]],
                 types = x$types, max = x$maximum, bounds = x$bounds,
                 verbose = control$verbose)
}

File: src/Rglpk_solve.c
/* This is the GLPK C Interface
 */

#include "Rglpk.h"
#include <stdio.h>
#include <setjmp.h>

// this is the solve function called from R
void R_glp_solve (int *lp_direction, int *lp_number_of_constraints,
                  int *lp_direction_of_constraints, double *lp_right_hand_side,
                  int *lp_number_of_objective_vars,
                  double *lp_objective_coefficients,
                  int *lp_objective_var_is_integer, 
                  int *lp_objective_var_is_binary,
                  int *lp_is_integer,                     //should be boolean
                  int *lp_number_of_values_in_constraint_matrix,
                  int *lp_constraint_matrix_i, int *lp_constraint_matrix_j,
                  double *lp_constraint_matrix_values,
                  int *lp_bounds_type, double *lp_bounds_lower,
                  double *lp_bounds_upper,
                  double *lp_optimum,
                  int *lp_col_stat,
                  double *lp_objective_vars_values,
                  double *lp_objective_dual_values,
                  int *lp_row_stat,
                  double *lp_row_prim_aux,
                  double *lp_row_dual_aux,
                  int *lp_verbosity,
                  int *lp_presolve,
                  int *lp_time_limit,
                  int *lp_status,
                  int *write_fmt,
                  char **fname,
                  int *write_sensitivity_report,
                  char **fname_sensitivity_report) {

  // GLPK problem object
  glp_prob *lp;
  // GLPK simplex control object
  glp_smcp control_sm;
  // GLPK mixed integer control object
  glp_iocp control_io;
  int i, kl, ku;
  jmp_buf env;

  // Patch provided by Xypron: A far jump is used to return if an
  // error occurs. Prior to that R crashed.
  if (setjmp(env)) {
    error("An error occured inside the GLPK library.");
  } else {
    glp_error_hook(Rglpk_error_hook, &env);

    // create problem object 
    lp = glp_create_prob();

    // Turn on/off Terminal Output
    if(*lp_verbosity==1)
      glp_term_out(GLP_ON);
    else
      glp_term_out(GLP_OFF);
    
    // direction of optimization
    if(*lp_direction==1)
      glp_set_obj_dir(lp, GLP_MAX);
    else
      glp_set_obj_dir(lp, GLP_MIN);
    
    // is it a mixed integer problem? -- seems to be an R glpk function
    //if(lp_integer)
    //lpx_set_class(lp, LPX_MIP);
    // add rows to the problem object
    if( *lp_number_of_constraints > 0 ){
      glp_add_rows(lp, *lp_number_of_constraints);
      for(i = 0; i < *lp_number_of_constraints; i++)
        switch(lp_direction_of_constraints[i]){
        case 1: 
          glp_set_row_bnds(lp, i+1, GLP_UP, 0.0, lp_right_hand_side[i]);
          break;
        case 2: 
          glp_set_row_bnds(lp, i+1, GLP_UP, 0.0, lp_right_hand_side[i]);
          break;
        case 3: 
          glp_set_row_bnds(lp, i+1, GLP_LO, lp_right_hand_side[i], 0.0);
          break;
        case 4: 
          glp_set_row_bnds(lp, i+1, GLP_LO, lp_right_hand_side[i], 0.0);
          break;
        case 5: 
          glp_set_row_bnds(lp, i+1, GLP_FX, lp_right_hand_side[i],
                           lp_right_hand_side[i]);
          break;
        }
    }
    
    // add columns to the problem object
    glp_add_cols(lp, *lp_number_of_objective_vars);
    kl = ku = 0;
    for(i = 0; i < *lp_number_of_objective_vars; i++) {
      glp_set_col_bnds(lp, i+1, lp_bounds_type[i], lp_bounds_lower[i], lp_bounds_upper[i]);
      // set objective coefficients and integer if necessary
      glp_set_obj_coef(lp, i+1, lp_objective_coefficients[i]);
      if (lp_objective_var_is_integer[i])
        glp_set_col_kind(lp, i+1, GLP_IV);
      if (lp_objective_var_is_binary[i])
        glp_set_col_kind(lp, i+1, GLP_BV);
    }
    // load the matrix
    // IMPORTANT: as glp_load_matrix requires triplets as vectors of the
    // form: ia[1] ... ia[n], we have to pass the pointer to the adress
    // [-1] of the corresponding vector 
    if( *lp_number_of_constraints > 0 ){
      glp_load_matrix(lp, *lp_number_of_values_in_constraint_matrix,
                      &lp_constraint_matrix_i[-1],
                      &lp_constraint_matrix_j[-1], &lp_constraint_matrix_values[-1]);
    }

    // write lp to file
    // mps_fixed := 1L, mps_free := 2L
    if ( *write_fmt > 0 ) {
        const char *out_name = fname[0];
        if ( *write_fmt < 3 ) {
            *lp_status = glp_write_mps(lp, *write_fmt, NULL, out_name);
        } else if ( *write_fmt == 3 ) {
            *lp_status = glp_write_lp(lp, NULL, out_name);
        } else {
            int future_flag = 0;
            *lp_status = glp_write_prob(lp, future_flag, out_name);
        }
        glp_delete_prob(lp);
        return;
    }
    
    // set optimizer control parameters
    glp_init_smcp(&control_sm);
    if (*lp_time_limit > 0) {
      control_sm.tm_lim = *lp_time_limit;
    }
    if (*lp_presolve == 1) {
      control_sm.presolve = GLP_ON;
    }
    
    // run simplex method to solve linear problem
    glp_simplex(lp, &control_sm);
    
    // retrieve status of optimization
    *lp_status = glp_get_status(lp);
    // retrieve optimum
    *lp_optimum = glp_get_obj_val(lp);
    // retrieve values of objective vars
    for(i = 0; i < *lp_number_of_objective_vars; i++) {
      lp_col_stat[i] = glp_get_col_stat(lp, i+1);
      lp_objective_vars_values[i] = glp_get_col_prim(lp, i+1);
      lp_objective_dual_values[i] = glp_get_col_dual(lp, i+1);
    }
    // retrieve primal/dual multipliers
    for(i = 0; i < *lp_number_of_constraints; i++) {
      lp_row_stat[i] = glp_get_row_stat(lp, i+1);
      lp_row_prim_aux[i] = glp_get_row_prim(lp, i+1);
      lp_row_dual_aux[i] = glp_get_row_dual(lp, i+1);
    }
    if(*lp_is_integer) {
      // set optimizer control parameters
      glp_init_iocp(&control_io);
      if (*lp_time_limit > 0) {
        control_io.tm_lim = *lp_time_limit;
      }
      if (*lp_presolve == 1) {
        control_io.presolve = GLP_ON;
      }
      // optimize
      glp_intopt(lp, &control_io);
      // retrieve status of optimization
      *lp_status = glp_mip_status(lp);
      
      // retrieve MIP optimum
      *lp_optimum = glp_mip_obj_val(lp);
      // retrieve MIP values of objective vars
      for(i = 0; i < *lp_number_of_objective_vars; i++){
        lp_objective_vars_values[i] = glp_mip_col_val(lp, i+1);
      }
      // retrieve MIP auxiliary variable values
      for(i = 0; i < *lp_number_of_constraints; i++) {
        lp_row_prim_aux[i] = glp_mip_row_val(lp, i+1);
      }
    }

    // write sensitivity analysis report
    if (*write_sensitivity_report == 1) {
      const char *out_name = fname_sensitivity_report[0];
      glp_print_ranges(lp, 0, NULL, 0, out_name);
    }

    // delete problem object
    glp_delete_prob(lp);
  }
}


